api/managementpb/ia/channels.proto

```diff
@@ -0,0 +1,172 @@
+syntax = "proto3";
+
+package ia.v1beta1;
+
+option go_package = "api/managementpb/ia;iav1beta1";
+
+import "github.com/mwitkow/go-proto-validators/validator.proto";
+import "google/api/annotations.proto";
+
+// TODO move it to parent directory / package once API is v1-stable
+
+// BasicAuth represents basic HTTP auth configuration.
+message BasicAuth {
+  string username = 1;
+  string password = 2;
+  string password_file = 3;
+}
+
+// TLSConfig represents TLS configuration.
+message TLSConfig {
+  string ca_file = 1;
+  string cert_file = 2;
+  string key_file = 3;
+  string server_name = 4;
+  bool insecure_skip_verify = 5;
+}
+
+// HTTPConfig represents HTTP client configuration.
+message HTTPConfig {
+  // Basic HTTP auth configuration. Zero value will remove it.
+  BasicAuth basic_auth = 1;
+  string bearer_token = 2;
+  string bearer_token_file = 3;
+  // TLS configuration. Zero value will remove it.
+  TLSConfig tls_config = 4;
+  string proxy_url = 5;
+}
+
+// EmailConfig represents email configuration.
+message EmailConfig {
+  bool send_resolved = 1;
+  string to = 2 [
+    (validator.field) = {
+      string_not_empty: true
+    }
+  ];
+}
+
+// SlackConfig represents Slack configuration.
+message SlackConfig {
+  bool send_resolved = 1;
+  string channel = 2 [
+    (validator.field) = {
+      string_not_empty: true
+    }
+  ];
+}
+
+// WebhookConfig represents webhook configuration.
+message WebhookConfig {
+  bool send_resolved = 1;
+  string url = 2 [
+    (validator.field) = {
+      string_not_empty: true
+    }
+  ];
+  HTTPConfig http_config = 3;
+  int32 max_alerts = 4;
+}
+
+// Channel represents a single Notification Channel.
+message Channel {
+  // reserved channels
+  // pagerduty_config = 3;
+  // pushover_config = 4;
+  // opsgenie_config = 6;
+  // victorops_config = 8;
+  // wechat_config = 9;
+
+  string channel_id = 1;
+  oneof channel {
+    EmailConfig email_config = 2;
+    SlackConfig slack_config = 5;
+    WebhookConfig webhook_config = 7;
+  }
+  // True if that channel is disabled.
+  bool disabled = 10;
+}
+
+message ListChannelsRequest {}
+
+message ListChannelsResponse {
+  repeated Channel channels = 1;
+}
+
+message AddChannelRequest {
+  string channel_id = 1 [
+    (validator.field) = {
+      string_not_empty: true
+    }
+  ];
+  // Email configuration. Exactly one config should be set.
+  EmailConfig email_config = 2;
+  // Slack configuration. Exactly one config should be set.
+  SlackConfig slack_config = 5;
+  // Webhook configuration. Exactly one config should be set.
+  WebhookConfig webhook_config = 7;
```

Comment:
A dedicated `type` field should be useful, I think. Now I should check every field to figure what type is received.

Reply:
We should check every field anyway to check that client sent only one field